### DISCUSSION TITLE
Preview of .pdf and .docx files

### DISCUSSION THREAD

**Discussion by mjmare** [wagtail/wagtail#10201] [General]
Is there a way to preview .pdf or .docx files in Wagtail and/or site powered by WT? 
This technically probably is not Wagtail specific, but maybe there is a plugin for this purpose?

The reason for asking is that Plone CMS offers a Document Preview plugin that I found very useful.

**Comment by thibaudcolas** [comment]
@mjmare what do you mean by "preview" in this context? As in a document viewer built into the page?

**Comment by mjmare** [comment]
Yes, that is what I would like to do: display the document's content within the page, possibly adorned with meta data (or other). 
So the user does not have to download the document, and open it in a separate application. 

(Oh, and while we are at it, I'd like to index the documents's content; but that merits its own topic.)

**Reply by zerolab** [reply]
fwiw, document indexing was covered by a third-party package - https://github.com/fourdigits/wagtail_textract which could use some love

**Comment by mjmare** [comment]
@zerolab Looks like a useful package! Thanks for the tip.

**Comment by tomdyson** [comment]
This plugin may be relevant too:

https://filepreviews.io/blog/2017/04/20/adding-document-previews-to-wagtail/

**Reply by mjmare** [reply]
Good to know!

**Comment by allcaps** [comment]
@mjmare Plone uses this package https://github.com/collective/collective.documentviewer 

It works like this:
1. Calls [Docsplit](https://github.com/collective/collective.documentviewer/blob/master/collective/documentviewer/convert.py#L464), a [command line util](https://documentcloud.github.io/docsplit/) to analyse the document and generate images.
2. Have a [some Javascript](https://github.com/collective/collective.documentviewer/blob/master/collective/documentviewer/resources/javascripts/viewer.js) to display the generated images and extracted data in a viewer.

I've migrated a couple of projects from Plone to Django/Wagtail and replicated this package.

Pros: 
- Docsplit will convert _any_ file type.
- Self hosted, no sensitive data to third-party servers.

Cons: 
- Docsplit is resource heavy and slow (had some memory leaks too). I'd had to use a task queue worker.
- Calling the Docsplit CLI feels like a hack and is very limited in specifying source, target locations, and filenames.

The primary reason to use a document viewer is that the user doesn't need to download the document. Keeping sensitive data on the server and off the users devices. If you don't have the no-download-because-sensitive-data-requirement, I would advice to offer a download an let the user open the document on their device, in their favorite viewer.

**Reply by mjmare** [reply]
My experience is that it makes perusing documents easier, if you (as a user) are not quite sure you have the richt document. When the user knows he has the right one, downloading is not a problem. 

BTW I'd be interested in your motivations and experiences moving from Plone to Wagtail. I found the default Plone UI to be very useful to get up and running. But not very easy to customise/extend. And major version upgrades were a pain. The handling of documents seemed more refined that the Wagtail way (AFAICT).
